BootloaderCorePkg/Include/Library/SocInitLib.h

```diff
@@ -7,6 +7,110 @@
+// 'RegInfo' (except value of reg) part is populated by the bootloader,
+// 'SmmBase' part is populated by the payload
+//
+
+      //  -------------------------- <----TSEG_BASE------
+      //  |     Signature (SCOM)   |                    ^
+      //  --------------------------                    |
+      //  |         CpuCnt         |                    |
+      //  --------------------------                    |
+      //  |         Unused         |                    |
+      //  --------------------------                    |
+      //  |          Rsvd          |                    |
+      //  --------------------------               SMMBASE_INFO
+      //  |       SmmBase[0]       |                    |
+      //  --------------------------                    |
+      //  |       SmmBase[1]       |                    |
+      //  --------------------------                    |
+      //  |         .....          |                    |
+      //  |         .....          |                    |
+      //  |         .....          |                    |
+      //  --------------------------                    |
+      //  |       SmmBase[n]       |                    V
+      //  -----------------------------------------------
+      //  |     Signature (S3SR)   |                    ^
+      //  --------------------------                    |
+      //  |         RegCnt         |                    |
+      //  --------------------------                    |
+      //  |       RegInfo[0]       |                    |
+      //  --------------------------                    |
+      //  |       RegInfo[1]       |            S3_SAVE_REG_INFO
+      //  --------------------------                    |
+      //  |         .....          |                    |
+      //  |         .....          |                    |
+      //  |         .....          |                    |
+      //  --------------------------                    |
+      //  |       RegInfo[n]       |                    V
+      //  -----------------------------------------------
+
+
+#pragma pack(1)
+
+typedef struct {
+  UINT8         Type;
+  UINT8         Width;
+  UINT8         Rsvd[2];
+  UINT32        Addr;
+  UINT32        Val;
+} REG_INFO;
+
+typedef struct {
+  UINT32        Signature;
+  UINT32        RegCnt;
+  REG_INFO      RegInfo[];
+} S3_SAVE_REG;
+
+#pragma pack()
+
+/**
+  This function uploads the register information in TSEG area
+  designated for S3 save/restore purpose. The 'Val' field of
+  the register information is updated by the payload.
+
+  @param    SmmBaseInfo     TSEG Base
+  @param    Type            IO/MEM register type to be saved
+  @param    Width           16-bit/32-bit wide register to be saved
+  @param    Addr            Addr of the register to be saved
+
+  @retval   None
+**/
+VOID
+EFIAPI
+S3SaveRegData (
+  IN  UINT8     *SmmBaseInfo,
+  IN  UINT8     Type,
+  IN  UINT8     Width,
+  IN  UINT32    Addr
+  );

```

Comment:
Prefer moving API and definition to S3SaveRestoreLib.h

Reply:
Will do.